File: python/sparklabs/batch/sales/dataset/distinct_customer.py
from sparklabs.batch.sales.dataset import utils
from sparklabs.batch.sales.dataset.models import Order, Product


def run() -> None:
    distinct_customer_count = utils.to_records(utils.orders, Order).map(
        lambda order: order.customer_id
    ).count()
    expensive_product = utils.to_records(utils.products, Product).takeOrdered(
        1, key=lambda product: -product.price
    )

    print(f"Total distinct customers: {distinct_customer_count}")
    print("Most Expensive Product:")
    print(expensive_product)


if __name__ == "__main__":
    run()

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataset/DistinctCustomer.scala
package io.devorbit.sparklabs.batch.sales.dataset

import io.devorbit.sparklabs.batch.sales.dataframe.Utils

object DistinctCustomer extends App {
  import Utils.sparkSession.implicits._

  private val distinctCustomerCount = DatasetUtils.orders.map(_.customer_id).count()
  private val expensiveProduct = DatasetUtils.products.orderBy(org.apache.spark.sql.functions.desc("price")).limit(1)

  println(s"Total distinct customers: $distinctCustomerCount")
  println("Most Expensive Product:")
  expensiveProduct.show(false)
}
